### PR TITLE
[WIP] add multi gpu support

### PR DESCRIPTION
**Pull request by boeddeker**
- Use the virtual mini batch size to distribute to different GPUs
 - Should work with most models without modification
 - Refactor the training loop to support multiple GPUs and
   - The iteration of the trainer can now be manipulated from hooks (e.g. back off)
   - The iteration indicates now the number of optimizer steps and not the number of example used
 - The step function signature changes to support different devices
 - The optimizer review is now separately reported to the hooks
 - Changed _maybe_add_loss_to_review to _review_to_loss_and_summary
 - Add log_error_state for better debugging
 - Remove Trainer.backward
 - Dropped backward in train_step
 - Add gradient check

 - rename train_iterator to train_dataset

 - [x] Support multiple GPUs to process examples from the virtual mini batch
 - [ ] Support multiple GPUs for validation

### CODE REVIEW

File: padertorch/contrib/examples/tasnet/train.py [hunk 1]
```diff
@@ -12,6 +12,8 @@
 from sacred import Experiment
 import sacred.commands
 
+# sacred.SETTINGS.CONFIG.READ_ONLY_CONFIG = False
+
```
Comment (jensheit) [inline padertorch/contrib/examples/tasnet/train.py:16]: Is this comment necessary?

File: padertorch/train/trainer.py [hunk 2]
```diff
@@ -248,7 +249,16 @@ def train(
         # Change model to train mode (e.g. activate dropout)
         self.model.train()
 
-        self.to(device)
+        if isinstance(device, (tuple, list)) \
+                and all([isinstance(d, int) for d in device]):
+            # multiple devises e.g. [0, 1], [0, 1, 2, 3], ...
```
Comment (jensheit) [inline padertorch/train/trainer.py:254]: the second if statement should be an assert in my opinion

File: padertorch/train/trainer.py [hunk 3]
```diff
@@ -266,48 +276,111 @@ def train(
...
+                            # torch.nn.parallel.DataParallel.
+                            # We also use the same functions
+                            # (i.e. replicate, parallel_apply and gather).
+                            #
+                            # The difference is, that we need no scatter,
+                            # because we simply use multiple examples and
+                            # the gather must only be applied on the loss.
+
+                            # Move copies of the model to each GPU
+                            replicas = replicate(self.model, device[:len(example)])
```
Comment (jensheit) [inline padertorch/train/trainer.py:342]: should we not do this before the virtual minibatch loop because now we are copying the same model to the devices in each iteration

File: padertorch/train/trainer.py (hunk 2 above)
Comment (boeddeker) [reply padertorch/train/trainer.py:254]: Yes, an assert produces a better exception msg. I will change it

File: padertorch/train/trainer.py (hunk 3 above)
Comment (boeddeker) [reply padertorch/train/trainer.py:342]: I copied the pattern from torch.nn.parallel.DataParallel.
They do the replicate in each forward call.
I haven't measured the overhead.
To keep the code simple, I would keep it for the moment here.

Comment (jensheit) [reply padertorch/train/trainer.py:342]: I think it makes sense to copy it in each forward call if you also always do a backwards step because the model is changing, but in our case we are just doing a backward step once for multiple forwards so we could copy it once

File: padertorch/contrib/examples/tasnet/train.py (hunk 1 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/train.py:16]: Not really. I added it, because I expected that the script fails without disable `READ_ONLY_CONFIG`.
Someone reported that the `betas` in the Adam optimizer caused some problems.
But this may only fail in some pytorch versions.

I will activate that line, to prevent some strange error messages from sacred.

**Comment by boeddeker** [comment]
I want to merge this PR and do the open points later:
- Support data parallel in validate
- Create the replicates only once after optimize
- Figure out which torch version have a working DataParallel (e.g. 1.4 doesn't work with RNNs)

Any comments to this PR?

I changed also the profiling results.
It is now simpler. The time per iteration is now measures explicit and all relative values are relative to the iteration time.

File: padertorch/contrib/examples/multi_gpu/train.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        trainer.model = DataParallel(trainer.model)
+    else:
+        print('#' * 79)
+        print('WARNING: Could not find multiple GPUs !!!')
+        print('#' * 79)
+    print('torch.cuda.device_count()', torch.cuda.device_count())
+    print('torch.cuda.is_available()', torch.cuda.is_available())
+
+    trainer.register_validation_hook(validation_iterator=dev_ds)
+    trainer.train(train_ds)
```
Comment (jensheit) [inline padertorch/contrib/examples/multi_gpu/train.py:164]: Shouldn't we add an example on how to use multiple gpus here instead of one showing what not to do?

File: padertorch/train/trainer.py [hunk 5]
```diff
@@ -362,39 +469,65 @@ def optimizer_zero_grad(self):
...
-    def validation_step(self, example):
-        return self.step(example, self.validate_timer)
+    def validation_step(self, model, example, device):
+        # [1:] -> ignore the loss. Is already in scalars.
+        return self.step(model, example, self.validate_timer, device)[1:]
 
-    def step(self, example, timer):
+    def step(self, model, example, timer, device):
         # TODO: Backup OutOfMemory
+        # with self.train_timer['time_per_step']:
```
Comment (thequilo) [inline padertorch/train/trainer.py:502]: Time per step is no longer reported?

File: padertorch/train/trainer.py [hunk 6]
```diff
@@ -362,39 +469,65 @@ def optimizer_zero_grad(self):
...
         with timer['time_per_to_device']:
-            example = self.model.example_to_device(example, self.device)
+            example = self.model.example_to_device(
+                example, device
+            )
```
Comment (thequilo) [inline padertorch/train/trainer.py:506]: Why reformatted? The original line is shorter than 79 chars.

Comment (thequilo) [reply padertorch/train/trainer.py:506]: Should this be `model` instead of `self.model`?

File: padertorch/contrib/examples/multi_gpu/train.py (hunk 4 above)
Comment (boeddeker) [reply padertorch/contrib/examples/multi_gpu/train.py:164]: Good point.
I created first this example, before I thought a build in support would be better,
I will change it to show both options.

File: padertorch/train/trainer.py [hunk 7]
```diff
@@ -257,57 +267,148 @@ def train(
...
+                            new_epoch = False
+                        elif minibatch_index == 0:
+                            # Call pre_step after getting the next example,
+                            # to correctly detect the next epoch
+                            with timer.pause():
+                                for hook in hooks:
+                                    hook.pre_step(self)
+
+                        if len(device) == 1:
+                            example, = example
```
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py:338]: This should never fail, this is just for the case that you change the trainer code.
But I will change the code to be more verbose (i.e. use your suggestion).

[the remark this answers is not in the document]
Comment (thequilo) [reply padertorch/train/trainer.py:338]: Yes, it "should" never fail. But you never know, maybe you at some point change the code before this line and it suddenly can fail

File: padertorch/train/trainer.py [hunk 8]
```diff
@@ -257,57 +267,148 @@ def train(
...
-                        )
+            train_iterable = None
+            while True:
+                new_epoch = False
+                if train_iterable is None:
+                    new_epoch = True
 
+                    # Call pre_step between the epochs.
+                    # We call it here, so it is done, before the iteration
+                    # over the train_dataset starts.
```
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py:310]: In general, it is not necessary.
The reason is, that this can reduce the memory footprint.
When you trigger the validation after an epoch the train dataset shouldn't start to load the next examples.

File: padertorch/train/trainer.py [hunk 9]
```diff
@@ -257,57 +267,148 @@ def train(
...
+                            # Release pytorch object to reduce memory footprint
+                            del example
+                            del model_output
+                            del review
+
+                            with self.train_timer['time_per_backward']:
+                                loss.backward(retain_graph=False)
+                            del loss
+
+                    # Only the summary hook will use optimizer_review
```
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py:403]: At the moment: yes.
I am also not completely happy with this solution.
Do you have another idea, how we can report the values from the optimizer?

File: padertorch/train/trainer.py [hunk 10]
```diff
@@ -257,57 +267,148 @@ def train(
...
+                    if optimize:
+                        with self.train_timer['time_per_optimize']:
+                            optimizer_review = self.optimizer_step()
+                            for hook in hooks:
+                                hook.post_step(self, None, None, optimizer_review)
```
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py:408]: Good idea. @JanekEbb and @jensheit what do you think?

File: padertorch/train/trainer.py (hunk 5 above)
Comment (boeddeker) [reply padertorch/train/trainer.py:502]: I dropped it, because it is relative similar to `time_per_iteration` ( = `time_per_data_loading` + `time_per_step`). Since backward is no longer in the step, it cannot directly be measured.

File: padertorch/train/trainer.py (hunk 6 above)
Comment (boeddeker) [reply padertorch/train/trainer.py:506]: Reformatted: This was in the merge, I didn't carefully check and used the previous formatting before you changed it. I will change it back.

model: It doesn't matter, but using `model` is better.

File: padertorch/train/trainer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jensheit) [reply padertorch/train/trainer.py:408]: sounds good to me, and the default would be that post_optimize just calls post_step?

File: padertorch/train/trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py:338]: As far as I know is `example, = example` a very cheap operation and if someone changes the code in the Trainer it is ok, to get a less informative error message (Better than silently fail).
But to be more verbose, I will add an assert.

File: padertorch/train/trainer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py:408]: I think the default should be that `post_optimize` does nothing.

[the remark this answers is not in the document]
Comment (jensheit) [reply padertorch/train/trainer.py:408]: If post_optimize does nothing then we have to add the behavior to the hooks which depent on the optimizer. In most cases this means calling the post_step or implement a similar behavior.

**Comment by boeddeker** [comment]
I have now a first number for the speedup of the DPRNN.
With 1 GPU the time per iteration is 0.74 and with 2 GPUs it is 0.4.
I think such a speedup is ok.

The replicate (i.e. copy the model to the second GPU) takes 3% of the execution time.
I think that is also ok.

File: padertorch/train/trainer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jensheit) [reply padertorch/train/trainer.py:408]: but you are right, the lines of code we are saving are not worth the possible confusing it may cause

**Comment by jensheit** [comment]
> I have now a first number for the speedup of the DPRNN.
> With 1 GPU the time per iteration is 0.74 and with 2 GPUs it is 0.4.
> I think such a speedup is ok.
> 
> The replicate (i.e. copy the model to the second GPU) takes 3% of the execution time.
> I think that is also ok.

Great, yes that looks like it is working.

File: padertorch/train/trainer.py (hunk 6 above)
Comment (boeddeker) [reply padertorch/train/trainer.py]: done

File: padertorch/train/trainer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py]: I pushed now a suggestion for `post_optimize`

File: padertorch/train/trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/train/trainer.py]: done

File: padertorch/train/trainer.py (hunk 2 above)
Comment (boeddeker) [reply padertorch/train/trainer.py:254]: done

File: padertorch/train/trainer.py (hunk 3 above)
Comment (boeddeker) [reply padertorch/train/trainer.py]: I will keep this point for another PR.
